### DISCUSSION TITLE
Database and Persistence

### DISCUSSION THREAD

**Discussion by tychoish** [tendermint/tendermint#6425] [Ideas]
_This is probably a draft ADR, but I think there are enough open questions that it would be useful to have a somewhat broader conversation before beginning that part of the process._

## Context
Tendermint has a very thin common wrapper that makes tendermint (largely) agnostic to the data storage layer (within the realm of the popular key-value/embedded databases.) This flexibility is not particularly useful: the benefits of a specific database engine in the context of tendermint is not particularly well understood, and the maintenance burden for multiple back ends is not commensurate with the benefit provided. Additionally, because the data storage layer is handled generically, and most tests run with an in-memory framework, it's difficult to take advantage of any higher-level features of a database engine.

Therefore we would like to replace the current persistence layer, with something that is both more simple and also gives us access to richer semantics. From a high level this project has the following goals:

- drop dependencies on storage engines that have a CGo dependency.
- encapsulate data format and data storage from higher-level services (e.g. reactors) within tendermint.
- select a storage engine that does not incur any additional operational complexity (e.g. database should be embedded.)
- provide databases semantics with sufficient ACID, snapshots, and transactional support.

Tendermint's data access patterns have the following characteristics:
- data size often exceeds memory.
- data is rarely mutated after it's written for most data (e.g. blocks), but small amounts of data are frequently mutated (e.g. peer information.)
- read patterns can be quite random.
- durable tranactionsal write operations would benefit application stability.

The following questions remain:
- what kind of data-access concurrency does tendermint require?
- would tendermint users SDK/etc. benefit from some shared database infrastructure?
- what are typical data sizes, and what kinds of memory sizes can we expect operators to be able to provide?
- in addition to simple persistence, what kind of additional semantics would tendermint like to enjoy (e.g. tranactionsal semantics, unique constraints, indexes, in-place-updates, etc.)?

Given the constraint of removing the CGo dependency, the decision is between "badger" and "boltdb" (in the form of the etcd/CoreOS fork,) as low level. On top of this and somewhat orthogonally, we must also decide on the interface to the database and how the larger application will have to interact with the database layer. Users of the data layer shouldn't ever need to interact with raw byte slices from the database, and should mostly have the experience of interacting with Go-types.

Badger is more consistently developed and has a broader featureset than Bolt. At the same time, Badger is likely more memory intensive and may have more overhead in terms of open file handles given it's model. At first glance, Badger is the obvious choice: it's actively developed and it has a lot of features that could be useful. Bolt is not without some benefits: it's stable and is maintained by the etcd folks, it's simpler model (single memory mapped file, etc,) may be easier to reason about.

I propose that we consider the following specific questions about storage engines:

- does Badger's evolving development, which may result in data file format changes in the future, and could restrict our access to using the latest version of the library between major upgrades, present a problem?
- do we do we have goals/concerns about memory footprint that Badger may prevent us from hitting, particularly as data sets grow over time?
- what kind of additional tooling might we need/like to build (dump/restore, etc.)?
- do we want to run unit/integration tests against a data files on disk rather than relying exclusively on the memory database?

This project will consist of the following aspects:

- selecting a storage engine, and modifying the tendermint code base to disallow any configuration of the storage engine outside of the code.
- remove the dependency on the current tm-db interfaces and replace with some internalized, safe, and ergonomic interface for data persistence with all required database semantics.
- update core tendermint code to use the new interface and data tools.

### Positive
- Removing CGo enables a great deal of simplification of the build, test, and release process.
- Reducing configuration options for data persistence will reduce operational complexity, test surface area, and make it easier to debug issues that interact with the data layer.
- Providing better ergonomics and safety at the persistence layer will make it easier to develop applications with reasonable data persistence patterns.
- Tighter integration with the storage layer may provide better options for optimization in the future.
### Negative
- Removing flexibility at the data layer may make it more difficult to work around operational problems or bugs with a specific storage engine.
- This impacts a large portion of the code, and will (necessarily) require a fully-resync/hard upgrade, while providing limited (if any) additional functionality to users.
### Neutral
N/A.
## Resources
- https://github.com/etcd-io/bbolt
- https://github.com/dgraph-io/badger
- https://dbdb.io/db/badgerdb
- https://dbdb.io/db/boltdb
- https://tech.townsourced.com/post/boltdb-vs-badger/
- https://github.com/timshannon/bolthold
- https://github.com/timshannon/badgerhold
- https://github.com/cockroachdb/pebble
- https://github.com/cosmos/cosmos-sdk/issues/7100 
- https://github.com/cosmos/cosmos-sdk/discussions/8297

**Comment by alexanderbez** [comment]
Excellent writeup @tychoish! I agree with most of the points laid out. In addition, I also believe Badger could be our best bet here. It's transactional/ACID compliant and supports snapshotting and it also supports high read and write throughput. In fact, it also has better sorted iteration latency.

Another good comparison: https://dgraph.io/blog/post/badger-lmdb-boltdb/

I would vote to just use Badger and call it a day. As for the interface that sits on top, I think we should also define that in this ADR, but we can keep it concise for now -- just the basics/CRUD. We can expand upon this interface in the future as we better understand user's needs.

**Reply by tac0turtle** [reply]
Nice write up!! I agree with going with one DB and calling it a day. Badger seems like the best candidate here. 

One thing I would like to add, that may be not prevalent right away, is disk size. There have been reports of badger increasing the size on disk by up to 4 times. This is without compression, most likely. Badger with compression brings in cgo i believe.

**Reply by tychoish** [reply]
> Another good comparison: https://dgraph.io/blog/post/badger-lmdb-boltdb/

This comparison is done by the folks who make Badger, so while it's interesting, I think it's useful to take a step back and take it in context with other information. 

> There have been reports of badger increasing the size on disk by up to 4 times. This is without compression, most likely. Badger with compression brings in cgo i believe.

This is interesting. The zstd compression library that badger uses depends on cgo. They also support snappy, and there's no reason that we can't use a gzip filter to compress our data, both of these options don't require cgo and probably help a bunch.

I'm not sure it's particularly possible to get a sense of "what grows by 4x" and if it's the block data or application data, and also if there may be other options to tune: I'm curious if badger ever releases disk space to the operating system after deletes, for example.

**Reply by tac0turtle** [reply]
Oh forgot about the other ones. 

goleveldb supports compression by default, badger I don't believe so. The comparison is between goleveldb as base point. I can ask the team(s) that reported this, we could also spin up some nodes this coming week to test differences.

**Comment by faddat** [comment]
Thank you so much for writing this up.

 I think that this is really a very important piece of commentary and I absolutely agree that we should avoid sego where possible and that we should standardize on a single way of storing data. Fewer choices make for happier programmers. In fact, I think that this will force us to ensure that the one that we do have works well and is Performant.

I can't think of any serious harm that will come from going down to just badger DB, but can think of many benefits.

**Reply by tychoish** [reply]
Thanks! I agree! I also think that if we put a reasonable interface/abstractions between the DB layer and our application(s?) that it'll make things easier to reason about and help improve the code all over. (and also make it possible to switch later if we need/want to.) 

Having spent more time than I'd like to admit working on database performance, I feel reasonably confident that the storage engine we use isn't going to be the bottleneck for tendermint, and that we should make a choice based on other factors like available database semantics, durability properties, and API ergonomics, rather than benchmarks.

**Reply by tessr** [reply]
> Having spent more time than I'd like to admit working on database performance, I feel reasonably confident that the storage engine we use isn't going to be the bottleneck for tendermint, and that we should make a choice based on other factors like available database semantics, durability properties, and API ergonomics, rather than benchmarks. 

💯 💯 :100:

**Comment by vorot93** [comment]
Hopefully not too late to the party.

We at [Erigon](https://github.com/ledgerwatch/erigon) started out as a fork of go-ethereum (with its LevelDB), then went all the way through Badger, Bolt and LMDB. We settled with [libmdbx](https://github.com/erthink/libmdbx), which is a well-maintained fork of LMDB and it provides us with everything needed: MVCC, extremely fast random access (basically in-memory sorted map) and sequential writes.

Here are the writeups detailing our journey:
https://github.com/ledgerwatch/erigon/wiki/Choice-of-storage-engine
https://github.com/ledgerwatch/erigon/wiki/Criteria-for-transitioning-from-Alpha-to-Beta#switch-from-lmdb-to-mdbx
